Update Cloud deployment upgrade best practices (#5703)
magento/devdocs: guides/v2.2/cloud/reference/discover-deploy.md

  - Deploy
---

> Build and deploy scripts activate every time you push code from your local workstation to the local environment using the local build process, or when you merge code to a remote Integration, Staging, or Production environment. These scripts generate new Magento code and provision configured services on the remote environment.

Reviewer: what do you mean under `local build process` ? 
I also would not say that those scripts `generate new Magento code`, probably you mean di compile process, but it is not a new code itself. anyway I think better to name it like a `prepare data, configurations and deploy Magento`.

> Build and deploy scripts activate every time you push code from your local workstation to the local environment using the local build process, or when you merge code to a remote Integration, Staging, or Production environment. These scripts generate new Magento code and provision configured services on the remote environment.

Reviewer: Also, build script not always may be run after push. But, not sure that we need to explain it here

The build and deploy process is slightly different for each plan:

-  **Starter plans**—For the Integration environment, every active branch builds and deploys to a full environment for access and testing. Fully test your code after merging to the `staging` branch. To go live, push `staging` to `master` to deploy to Production. You have full access to all branches through the Project Web Interface and the CLI commands.

> -  **Pro plans**—For the Integration environment, every _active_ branch builds and deploys to a full environment for access and testing. You must merge your code to the `integration` branch before you can merge to the Staging environment and then the Production environment. You can only merge to Staging and Production using CLI commands with SSH or using the Project Web Interface.

Reviewer 2: The last sentence is confusing. Maybe "You can merge to the Staging and Production environments using the Project Web Interface or you can log in to the remote server using SSH and merge them from the command line."

## Track the process {#track}

> You can track build and deploy actions in real-time using the terminal or the Project Web Interface. Status messages—`in-progress`, `pending`, `success`, or `failed`—display during the deployment process. You can view details in the log files. See [Log locations]({{ page.baseurl }}/cloud/project/log-locations.html).

Reviewer: deploy logs are not displayed on integration environment, only build and post-deploy

> You can track build and deploy actions in real-time using the terminal or the Project Web Interface. During the deployment process, the following status messages display on the command line, or in the interface: `in-progress`, `pending`, `success`, or `failed`. You can view details in the log files. See [Log locations]({{ page.baseurl }}/cloud/project/log-locations.html).

Reviewer 2: When I see "following" I think of a list or something that follows the paragraph, not the end of the sentence.

```suggestion
You can track build and deploy actions in real-time using the terminal or the Project Web Interface. Status messages—`in-progress`, `pending`, `success`, or `failed`—appear during the deployment process. You can view details in the log files. See [Log locations]({{ page.baseurl }}/cloud/project/log-locations.html).
```

-  **Pro plans**—For the Integration environment, every _active_ branch builds and deploys to a full environment for access and testing. You must merge your code to the `integration` branch before you can merge to the Staging environment and then the Production environment. You can merge to the Staging and Production environments using the Project Web Interface or using SSH and Magento Cloud CLI commands.

## Track the process {#track}

You can track build and deploy actions in real-time using the terminal or the Project Web Interface. Status messages—`in-progress`, `pending`, `success`, or `failed`—display during the deployment process. You can view details in the log files. See [Log locations]({{ page.baseurl }}/cloud/project/log-locations.html).

If you are using external GitHub repositories, the log of the operations does not display in the GitHub session. You can still follow activity in the interface for the external repository and in the {{site.data.var.ece}} Project Web Interface. See [Integrations]({{ page.baseurl }}/cloud/integrations/cloud-integrations.html).

{:.bs-callout-info}

> Deploy logs are available only for Staging and Production environments. See [Log locations]({{ page.baseurl }}/cloud/project/log-locations.html).

Reviewer: I apologize, probably my previous comment was not very clear

> deploy logs are not displayed on integration environment, only build and post-deploy

I meant that you cannot see Deploy logs in the Cloud UI for integration environments. But in our file logs you can see logs for every phase, no matter what environment you use.

> Deploy logs are available only for Staging and Production environments. In Integration environments, you can view only build and post-deploy logs.

Reviewer 2: This is not completely true. On Integration environments, you have the `cloud.log` file, which shows start and stop messages for each hook.

```suggestion
Deploy logs are available only for Staging and Production environments. See [Log locations]({{ page.baseurl }}/cloud/project/log-locations.html).
```

-  [`app/etc/config.php`]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html)—defines the [system-specific settings]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html#cloud-clp-settings) Magento uses to configure your store. Magento auto-generates this file if it does not detect it during the build phase and includes a list of modules and extensions. If the file exists, the build phase continues as normal, compresses static files using `gzip`, and deploys the files. If you follow [Configuration Management]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html) at a later time, the commands update the file without requiring additional steps.

  {%include cloud/note-static-content-deploy-config.md %}

## Required files for your Git branch {#requiredfiles}

Your Git branch must have the following files for building and deploying in your local environment and to Integration, Staging, and Production environments:

> -  `auth.json`—in the root Magento directory. This file includes the Magento Authentication keys entered when creating the project. The file is generated as part of autoprovisioning a new project using a blank template. If you need to verify the file and settings, see [Troubleshooting]({{ page.baseurl }}/cloud/trouble/troubleshoot-deployment.html).

Reviewer 2: Why remove "deployment"? That is the name of the topic...unless you want them to relate more to the entire troubleshooting section.

-  [`app/etc/config.php`]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html)—auto-generates during the build phase if it does not exist.
-  [`.magento.app.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_magento-app.html)—updates and saves to the root directory.
-  [`.magento/services.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_services.html)—updates and saves to the `magento/` directory.
-  [`.magento/routes.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_routes.html)—updates and saves to the `magento/` directory.

## Best practices for builds and deployment {#best-practices}

We highly recommend the following best practices and considerations for your deployment process:

-  **Ensure that you are running the most current version of ece-tools**–See {{ site.data.var.ct }} [Release notes]({{ page.baseurl }}/cloud/release-notes/cloud-tools.html).

> -  **Follow the build and deploy process**–Ensure that you have the correct code in each environment to avoid overwriting configuration when merging code between environments. For example, to make configuration changes that apply to all environments, modify and test the changes in the local environment before deploying to Integration, and then deploy and test the changes in Staging before deploying to Production. When you merge from one environment to another, the deployment overwrites all code in the remote environment, except environment-specific configuration and settings. See [Build and deploy full steps](#steps).

Reviewer 2: Avoid overwriting configurations (with an s) or a configuration...something seems missing.

> -  **Use the same variables across environments**–The values for these variables may differ across environments, but the same variables are usually required in each environment. See [Configuration management for store settings](/{{ page.baseurl }}//cloud/live/sens-data-over.html).

Reviewer 2: The word usually is splitting the verb

-  **Keep sensitive configuration values and data in environment-specific variables**–This includes variables specified using the Magento Cloud CLI, the Project Web interface, or added to the `env.php` file. See [Working with environment variables]({{ page.baseurl }}/cloud/env/working-with-variables.html).

-  **Ensure that all code is available in the environment branch**–If you point to or include hooks to code in other branches, such as a private branch, you may encounter problems with the build and deploy process. For example, if you include a theme from a private repository in your branch, the theme will not build with the Magento code.

-  **Add new extensions, integrations, and code in iterated branches**–Make and test changes locally, push to Integration, then to Staging and Production. Test and resolve issues in each environment before merging the updates to the next environment. Some extensions and integrations must be enabled and configured in a specific order due to dependencies. Adding these in groups can make your build and deploy process much easier and help determine where issues occur.

> -  **Verify service versions and relationships and ability to connect as needed**–See [Service versions]({{ page.baseurl }}/cloud/project/project-conf-files_services.html#service-versions) and [Service relationships]({{ page.baseurl }}/cloud/project/project-conf-files_services.html#service-relationships).

Reviewer 2: This may need a little more help.
```suggestion
-  **Verify service versions and relationships and the ability to connect**–Verify the services that are available to your application and ensure you are using the most appropriate version. See [Service versions]({{ page.baseurl }}/cloud/project/project-conf-files_services.html#service-versions) and [Service relationships]({{ page.baseurl }}/cloud/project/project-conf-files_services.html#service-relationships).
```

-  **Check Production environment configuration**–Before deploying to Production, complete the following tasks:

   -  Ensure that you can connect to all three nodes in the Production environment using [SSH]({{ page.baseurl }}/cloud/env/environments-ssh.html#ssh).

   -  Verify that Indexers are set to _Update on Schedule_. See [Indexing modes]({{ page.baseurl }}/extension-dev-guide/indexing.html) in the _Extension Developer Guide_.

   -  Run the {{ site.data.var.ece }} [smart wizard]({{ page.baseurl }}/cloud/deploy/smart-wizards.html) to ensure that your Cloud project configuration follows best practices for build and deployment.

>    -  Prepare the environment by updating any environment-specific variables in the Production code, verifying service availability and compatibility, and making any other required configuration changes.

Reviewer 2: A lot of these seem like they are already pointed out earlier, but maybe it's okay to say them again.

Author: I think the issue is that people are not checking the Production environment before they deploy, so this is a reminder.

   -  Review the {{ site.data.var.ee }} release notes and extension documentation to ensure that you have implemented any workarounds or configuration changes required to address known functional issues and bugs related to the upgraded Magento version and extensions.

   -  Check service configuration and version compatibility and upgrade services as needed. See [Services]({{ page.baseurl }}/cloud/project/project-conf-files_services.html).

-  **Plan and perform a successful upgrade on local and Integration environments, before deploying to Staging**–After upgrade, test your deployment and resolve any issues. See [Build and deploy on local]({{ page.baseurl }}/cloud/live/live-sanity-check.html).

-  **Monitor the upgrade process**–When you deploy to an environment, review the deploy process status messages and mitigate issues as needed. Review the Cloud [logs]({{page.baseurl}}/cloud/project/log-locations.html) for detailed log messages.

-  **Merge code to Staging, and then to Production**–After the deploy completes, test and resolve any issues in the Staging environment before pushing changes to the Production environment.

>    -  Use SSH to verify availability of the MySQL service in remote environments. See [Set up MySQL]({{ page.baseurl }}/cloud/project/project-conf-files_services-mysql.html).

Reviewer 2: ```suggestion
   -  Use SSH to log in to the remote server and verify availability of the MySQL service in remote environments. See [Set up MySQL]({{ page.baseurl }}/cloud/project/project-conf-files_services-mysql.html).
```

   -  [Backup the database]({{ page.baseurl}}/cloud/project/project-upgrade.html#back-up-the-database).

   -  [Verify available storage in the remote environment]({{ page.baseurl }}/cloud/project/manage-disk-space.html).

      -  Check the size of the upgraded database and verify that the `services.yaml` file has enough disk space allocated.

      -  Free up space in `/log` and `/tmp` directories before deploying. See [Log locations]({{page.baseurl}}/cloud/project/log-locations.html).

   -  Make any required updates to environment-specific settings before [deploying to the remote environment]({{ page.baseurl }}/cloud/live/stage-prod-migrate.html).

   -  Merge updated code to deploy to the remote environment.

-  **Complete Post upgrade tasks**–

>    -  Use SSH to connect to the remote environment and verify the following:

Reviewer 2: ```suggestion
   -  Use SSH to log in to the remote server and verify the following:
```

-  [`.magento/services.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_services.html)—defines the services Magento uses by name and version. For example, this file may include versions of MySQL, PHP extensions, and Elasticsearch. These are referred to as *services*.
-  [`app/etc/config.php`]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html)—defines the [system-specific settings]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html#cloud-clp-settings) Magento uses to configure your store. Magento auto-generates this file if it does not detect it during the build phase and includes a list of modules and extensions. If the file exists, the build phase continues as normal, compresses static files using `gzip`, and deploys the files. If you follow [Configuration Management]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html) at a later time, the commands update the file without requiring additional steps.

## Required files for your Git branch {#requiredfiles}

Your Git branch must have the following files for building and deploying in your local environment and to Integration, Staging, and Production environments:

> -  `auth.json`—in the root Magento directory. This file includes the Magento Authentication keys entered when creating the project. The file is generated as part of autoprovisioning a new project using a blank template. If you need to verify the file and settings, see [Troubleshooting]({{ page.baseurl }}/cloud/trouble/troubleshoot-deployment.html).

Author: ```suggestion
-  `auth.json`—in the root Magento directory. This file includes the Magento Authentication keys entered when creating the project. The file is generated as part of autoprovisioning a new project using a blank template. If you need to verify the file and settings, see [Troubleshooting deployment]({{ page.baseurl }}/cloud/trouble/troubleshoot-deployment.html).
```

-  [`app/etc/config.php`]({{ site.baseurl }}/guides/v2.2/cloud/live/sens-data-over.html)—auto-generates during the build phase if it does not exist.
-  [`.magento.app.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_magento-app.html)—updates and saves to the root directory.
-  [`.magento/services.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_services.html)—updates and saves to the `magento/` directory.
-  [`.magento/routes.yaml`]({{ page.baseurl }}/cloud/project/project-conf-files_routes.html)—updates and saves to the `magento/` directory.

## Best practices for builds and deployment {#best-practices}

We highly recommend the following best practices and considerations for your deployment process:

-  **Ensure that you are running the most current version of the `{{site.data.var.ct}}` package**–See [Release notes for {{site.data.var.ct}}]({{ page.baseurl }}/cloud/release-notes/cloud-tools.html).

> -  **Follow the build and deploy process**–Ensure that you have the correct code in each environment to avoid overwriting configuration when merging code between environments. For example, to make configuration changes that apply to all environments, modify and test the changes in the local environment before deploying to Integration, and then deploy and test the changes in Staging before deploying to Production. When you merge from one environment to another, the deployment overwrites all code in the remote environment, except environment-specific configuration and settings. See [Build and deploy full steps](#steps).

Reviewer 2: ```suggestion
-  **Follow the build and deploy process**–Ensure that you have the correct code in each environment to avoid overwriting a configuration when merging code between environments. For example, to make configuration changes that apply to all environments, modify and test the changes in the local environment before deploying to Integration, and then deploy and test the changes in Staging before deploying to Production. When you merge from one environment to another, the deployment overwrites all code in the remote environment, except environment-specific configuration and settings. See [Build and deploy full steps](#steps).
```

> -  **Follow the build and deploy process**–Ensure that you have the correct code in each environment to avoid overwriting configuration when merging code between environments. For example, to make configuration changes that apply to all environments, modify and test the changes in the local environment before deploying to Integration, and then deploy and test the changes in Staging before deploying to Production. When you merge from one environment to another, the deployment overwrites all code in the remote environment, except environment-specific configuration and settings. See [Build and deploy full steps](#steps).

Author: ```suggestion
-  **Follow the build and deploy process**–Ensure that you have the correct code in each environment to avoid overwriting configurations when merging code between environments. For example, to make configuration changes that apply to all environments, modify and test the changes in the local environment before deploying to Integration, and then deploy and test the changes in Staging before deploying to Production. When you merge from one environment to another, the deployment overwrites all code in the remote environment, except environment-specific configuration and settings. See [Build and deploy full steps](#steps).
```
